### PR TITLE
Added admin create environment as github action

### PR DESCRIPTION
**Pull request by Dee-pak**
Exposed admin create verb as github action.

### CODE REVIEW

File: create-environment/action.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+  templates:
+    description: 'The Dynamics 365 apps that needs to be deployed to the environment. Passed as comma separated values.'
+    requried: false
+
+  domain:
+    description: 'The domain name is of the environment URL. Eg: https://{contoso}0.crm.dynamics.com.'
+    requried: false
+
+
```
Comment (swapnilbhalgat) [inline create-environment/action.yml:46]: Consecutive blank lines

File: create-environment/action.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  password-secret:
+    description: 'Power Platform password. Do NOT checkin password, instead create a secret and reference it here with: see: https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets#using-encrypted-secrets-in-a-workflow'
+    required: true
+
+  name:
+    description: 'The name of the environment you are creating.'
+    required: true
+
+  region:
+    description: 'The name of the region for your environment. Default is 'unitedstates'.'
```
Comment (swapnilbhalgat) [inline create-environment/action.yml:23]: you can also specify default values as default: 'unitedstates'

File: .github/workflows/CI.yml [hunk 3]
```diff
@@ -86,3 +86,13 @@ jobs:
         environment-url: 'https://davidjenD365-1.crm.dynamics.com'
         user-name: 'davidjen@davidjenD365.onmicrosoft.com'
         password-secret: ${{ secrets.password }}
+
+    - name: Test create-environment action
```
Comment (davidjenni) [inline .github/workflows/CI.yml:90]: remove this action from the CI/PR workflow

File: create-environment/action.yml (hunk 2 above)
Comment (davidjenni) [reply create-environment/action.yml:23]: not just a nit: make sure the action parameter defaults here agree with what index.ts does. This is indeed a bit of double book keeping; action.yaml also serves as the public documentation of the action

File: create-environment/action.yml [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+name: 'create-environment'
+description: 'Power Platform Admin Create Environment'
+inputs:
+  environment-url:
```
Comment (davidjenni) [inline create-environment/action.yml:6]: uhm, no: that can't work for the admin call... there is no url yet before it is created. The envUrl is a result of the creation.

File: src/actions/create-environment/index.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        if (!password || password.length === 0) {
+            return core.setFailed('Missing password! Specify one by setting input: \'password-secret\'');
+        }
+
+        const envName = core.getInput('name', { required: true});
+        const envRegion = core.getInput('region', {required: true});
+        const envType = core.getInput('type', {required: true});
+
+        const pac = factory.getRunner('pac', process.cwd());
+        await pac.run(['auth', 'create', '--url', envUrl, '--username', username, '--password', password]);
```
Comment (davidjenni) [inline src/actions/create-environment/index.ts:29]: that code won't work: for CRUD operations, this needs to be a pac auth create --kind ADMIN

File: src/actions/create-environment/index.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        const username = core.getInput('user-name', { required: true });
+        core.info(`environmentUrl: ${envUrl}; login as user: ${username}`);
+
+        const password = core.getInput('password-secret', { required: true });
+        if (!password || password.length === 0) {
+            return core.setFailed('Missing password! Specify one by setting input: \'password-secret\'');
+        }
+
+        const envName = core.getInput('name', { required: true});
+        const envRegion = core.getInput('region', {required: true});
```
Comment (davidjenni) [inline src/actions/create-environment/index.ts:25]: disconnect between action.yml and code: here both are required, but action.yml has no default values

**Review by davidjenni — CHANGES_REQUESTED** [review]

**Review by swapnilbhalgat — APPROVED** [review]

File: create-environment/action.yml [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    default: 'unitedstates'
+
+  type:
+    description: 'The type of environment (Trial, Sandbox, Production, SubscriptionBasedTrial).'
+    required: true
+    default: 'Sandbox'
+
+  currency:
+    description: 'The currency to use for the environment.'
+    requried: false
```
Comment (davidjenni) [inline create-environment/action.yml:31]: typo!

Comment (davidjenni) [reply create-environment/action.yml:31]: fix the remaining 3 properies as well, all same copy/paste error

File: src/actions/create-environment/index.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        const envType = core.getInput('type', {required: true});
+
+        const pac = factory.getRunner('pac', process.cwd());
+        await pac.run(['auth', 'create', '--kind', 'ADMIN', '--username', username, '--password', password]);
```
Comment (davidjenni) [inline src/actions/create-environment/index.ts:26]: do add a pac auth clear to wipe away existing auth profiles

**Review by davidjenni — CHANGES_REQUESTED** [review]

File: create-environment/action.yml [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+  domain:
+    description: 'The domain name is of the environment URL. Eg: https://{contoso}0.crm.dynamics.com.'
+    requried: false
+
+runs:
+  using: 'node12'
+  main: '../dist/actions/create-environment/index.js'
+  # main: '../out/actions/create-environment/index.js'
```
Comment (davidjenni) [inline create-environment/action.yml:50]: note:  once all is working, fixed and you have a passing PR loop, you'll need to locally run 'npm run dist' to create an additional commit to update the dist folder with the added actions...

**Review by davidjenni — CHANGES_REQUESTED** [review]

File: create-environment/action.yml [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    required: true
+
+  password-secret:
+    description: 'Power Platform password. Do NOT checkin password, instead create a secret and reference it here with: see: https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets#using-encrypted-secrets-in-a-workflow'
+    required: true
+
+  name:
+    description: 'The name of the environment you are creating.'
+    required: true
+    default: 'newEnvironment'
```
Comment (davidjenni) [inline create-environment/action.yml:17]: not a good default we can choose; just make it required, no default

File: create-environment/action.yml [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    required: false
+    default: 'USD'
+
+  language:
+    description: 'The language to use for the environment.'
+    required: false
+    default: 'English'
+
+  templates:
+    description: 'The Dynamics 365 apps that needs to be deployed to the environment. Passed as comma separated values.'
```
Comment (davidjenni) [inline create-environment/action.yml:40]: typo/grammar: "The Dynamics 365 that need to be deployed to...."

File: dist/actions/clone-solution/index.js [hunk 12]
```diff
@@ -16,6 +16,7 @@ var __importStar = (this && this.__importStar) || function (mod) {
 };
 Object.defineProperty(exports, "__esModule", ({ value: true }));
 const os = __importStar(__webpack_require__(2087));
+const utils_1 = __webpack_require__(5278);
```
Comment (davidjenni) [inline dist/actions/clone-solution/index.js:19]: sad how much noise webpack is adding... oh well

**Review by davidjenni — APPROVED** [review]

**Pull request merged**